### PR TITLE
[TST] 1D-Poisson test added

### PR DESCRIPTION
**Pull request by gusespinola**
A 1D-Poisson test was added, also we use an old version of PD-GMRES for testing.

### CODE REVIEW

File: tests/test_pd_GMRES.m [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    try % assignment of 'localfunctions' is necessary in Matlab >= 2016
+        test_functions = localfunctions(); %#ok<*NASGU>
+    catch % no problem; early Matlab versions can use initTestSuite fine
+    end
+
+    initTestSuite;
+
+end
+
+% function test_pd_GMRES_0()
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:15]: Please remove commented lines of code.

File: tests/test_pd_GMRES.m [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    % ============
+    % Solve a 1D-Poisson equation at [aStart, aEnd] = [0, 1]
+    % - u''(x) = f(x)
+    % subject to
+    % u(0) = g(0) = 0, u(1) = g(1) = 1;
+    % where
+    % f(x) = 6*x; g(x) = x^3
+    % Discretization results in a linear system Au = b
+    
+    % Discretization grid
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:65]: `f` and `g` are independent of the discretisation grid. Consider placing them above L65, with a header comment such as `% Source term and boundary values`.

File: tests/test_pd_GMRES.m [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    g = @(x) x.^3;
+    NODES = 9;
+    INNERNODES = NODES-2;
+    aStart = 0;
+    aEnd = 1;
+    h = ( aEnd - aStart ) / (NODES - 1);
+    
+    % Left-hand side matrix 'A'
+    A = 2*eye(INNERNODES) - diag(ones(INNERNODES-1,1), 1) - diag(ones(INNERNODES-1, 1), -1);
+    %A(1, 1:2) = [1 0];x
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:76]: Please remove commented lines of code.

File: tests/test_pd_GMRES.m [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    %A(NODES, NODES-1:NODES) = [0 1];
+    
+    % Right-hand side 'b'
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:79]: We can explain briefly that the RHS is composed by the contributions from the boundary conditions `g(x)` and the source term `f(x)`.

File: tests/test_pd_GMRES.m [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    b = h*h*f( linspace(aStart + h, aEnd - h, INNERNODES) )'; % b = h^2*f(x)
+    % Dirichlet conditions are here
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:81]:
```suggestion
    % Add contribution from Dirichlet nodes to vector 'b'
```

File: tests/test_pd_GMRES.m [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    % b(1,1) = g( aStart );
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:82]: The contribution here is zero, but I will keep it anyways. This will allow us to later change to a non-zero Dirichlet bc (if we need it for some reason).

File: tests/test_pd_GMRES.m [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    b(INNERNODES,1) = b(INNERNODES,1) + g( aEnd );
+    
+    % Exact solution 'uExact'
+    uExact = g( linspace(aStart + h, aEnd - h, INNERNODES) )'
+    
+    % CALL ALGORITHMS: PD_GMRES, ADAPTIVE_GMRES, SWITCH_GMRES, etc.
+    % PD_GMRES(m)
+    % The original idea was to compute the average execution time,
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:90]: We won't be measuring average execution time (certainly not in this test). Please remove commented lines 90 - 95.

File: tests/test_pd_GMRES.m [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    % we may discuss if this is still necessary 
+    %rootFolder = fileparts(pwd); % go to the root folder
+    %srcFolder = fullfile(rootFolder, 'src') % enter the data folder
+    %cd(srcFolder)
+    % [time, logres_pd_gmres]=pd_GMRES(A,b, mPD, alpha, delta,itermax); 
+    % Parameters for PD_GMRES
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:96]: I see that there are non-default parameters being used by PD_GMRES. Is there a rationale for this? Can we use default parameters? Maybe you can comment on this @jccf19

File: tests/test_pd_GMRES.m [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    alpha=-3; %2
+    delta=5; %0.8
+    opts_tol=1e-9;
+    itermax=1000;
+    p = 1;
+    for i=1:p
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:102]: Plotting-related functionality should not be part of the test.

File: tests/test_pd_GMRES.m [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        color_pd_gmres='b';
+        mPD=3;
+        %alpha=2;
+        %rootFolder = fileparts(pwd); % go to the root folder
+        %srcFolder = fullfile(rootFolder, 'src'); % enter the data folder
+        %cd(srcFolder)
+        uBackslash  = A\b;
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:109]: We do not need to solve the system with backslash, since we know we exact solution.

File: tests/test_pd_GMRES.m [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        [time, logres_pd_gmres, uPD_GMRES]=pd_GMRES(A,b, mPD, alpha, delta,itermax,opts_tol);
+        uPD = uPD_GMRES;
+        uExact = g( linspace(aStart + h, aEnd - h, INNERNODES) )';
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:112]: It looks like `uPD` can be calculated once outside the loop. Right?

File: tests/test_pd_GMRES.m [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        %sol7(size(sol7,1)+1,:)= [time];
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:113]: Remove comment.

File: tests/test_pd_GMRES.m [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    end
+    
+    % assert test
```
Comment (jhabriel) [inline tests/test_pd_GMRES.m:116]: Remove commented lines.

**Review by jhabriel — CHANGES_REQUESTED** [review]
I think we are in a decent place to start. 

A few general comments though:

1. Commented code (not comments) must be removed before making a pull request.
2. Plotting functionality should not be included as part of the algorithm nor the tests. 
3. Comments should be written only in English, as we agreed on.

@gusespinola: Let me know if you want to discuss something in person.

File: src/pd_GMRES.m [hunk 14]
```diff
@@ -1,5 +1,4 @@
-function [tiempo_C, log_res] = pd_GMRES(A, b, m_PD, alpha, delta, ...
-                                           iter_max)
+function [tiempoC, logres, xx]=pd_GMRES(A,b, mPD, alpha, delta,itermax,tol)
```
[the remark this answers is not in the document]
Comment (gusespinola) [reply src/pd_GMRES.m:1]: Added description for 'xx' - the linear system solution. Names for the input/output arguments may change.

File: tests/test_pd_GMRES.m (hunk 11 above)
Comment (gusespinola) [reply tests/test_pd_GMRES.m:64]: It was used only for debugging, now it is deleted.

File: src/pd_GMRES.m [hunk 15]
```diff
@@ -44,32 +43,39 @@
...
+mmax=n-1; %se puede considerar que no tiene cota superior, antes de las 1000 iteraciones  no logra alcanzar mmax con alpha=-3 y delta=5
+mstep=1;
+maxit=itermax;
+%maxit=1000;
 
 %Parameters 
-alpha_0 = alpha; % BORRAR?
-delta_0 = delta; % BORRAR?
+alpha0=alpha;
+delta0=delta;
```
[the remark this answers is not in the document]
Comment (gusespinola) [reply src/pd_GMRES.m:57]: 'alpha' and 'delta' are the proportional and derivative coefficients of the PD controller for restart parameter 'm'.

According to reference, we rather should change to 'alphaD' and 'alphaD' respectively.

File: tests/test_pd_GMRES.m (hunk 8 above)
Comment (gusespinola) [reply tests/test_pd_GMRES.m:53]: Sure! We can see the reference for more details.

File: src/pd_GMRES.m [hunk 16]
```diff
@@ -44,32 +43,39 @@
...
+
 
-flag = 0; % FLAG DE CONVERGENCIA?
-% convergencia = 0; % BORRAR?
 
+flag=0;
```
[the remark this answers is not in the document]
Comment (gusespinola) [reply src/pd_GMRES.m:58]: flag for convergence, now limited to 
-1 for convergence (relres<tol), 
-0 for stagnation/no convergence before the max number of cycles

MATLAB offers other values for this flag, we can check again this feature.

File: tests/test_pd_GMRES.asv
[the remark this answers is not in the document]
Comment (gusespinola) [reply tests/test_pd_GMRES.asv:1]: .asv, AutoSave file, I created a .gitignore for these type of files.

File: src/pd_GMRES.m [hunk 17]
```diff
@@ -20,170 +17,128 @@
...
-alpha_0 = alpha; % BORRAR?
-delta_0 = delta; % BORRAR?
-
-flag = 0; % FLAG DE CONVERGENCIA?
-% convergencia = 0; % BORRAR?
-
+x0 = zeros(size(b,1),1); % x0: First guess for solution vector 'x' 
+mInitial = mPD;
+mmin = 1;
+mmax = n-1; %It can be considered that it does not have an upper bound, before 1000 iterations it cannot reach mmax with alphaP=-3 and alphaD=5
```
Comment (jccf19) [inline src/pd_GMRES.m:53]: Please remove this comment, no longer needed.

File: tests/test_pd_GMRES.m [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    % we may discuss if this is still necessary 
+
+    % Parameters for PD_GMRES
+    alpha=-3;
+    delta=5;
+    opts_tol=1e-9;
+    itermax=1000;
+    p = 1;
+    for i=1:p
+        color_pd_gmres='b';
```
Comment (jccf19) [inline tests/test_pd_GMRES.m:64]: Please remove this line, there is no need.

**Review by jhabriel — APPROVED** [review]
Great job @gusespinola and thanks for the effort! This is a considerable improvement from the first version.

I will start working on #14 as soon as this PR is merged.
